File: cprogram/forExam/linkedList.c
#include<stdio.h>
#include<stdlib.h>

struct node{
	int data;
	struct node* next;
};

struct node* createNode(int data){
	struct node* newNode=(struct node*)malloc(sizeof(struct node));
	newNode->data=data;
	newNode->next=NULL;

	return newNode;
}

void insertAtBeg(struct node** head,int data){
	struct node* newNode=createNode(data);
	newNode->next=*head;
	*head=newNode;

}

void insertAtEnd(struct node** head,int data){
	struct node* newNode=createNode(data);
	if(*head==NULL){
		*head=newNode;
		return;
	}
	struct node* temp=*head;
	while(temp->next!=NULL){
		temp=temp->next;
	}
	temp->next=newNode;
}
int search(struct node* head,int key){
	struct node* temp=head;
	while(temp!=NULL){
		if(key==temp->data){
			return 1;
		}
			temp=temp->next;
	}
	return 0;
}	

void deleteNode(struct node** head,int key){
	struct node* temp=*head;
	struct node* prev=NULL;

	if(temp!=NULL&&temp->data==key){
		*head=temp->next;
		free(temp);
		return;
	}

	while(temp!=NULL&&temp->data!=key){
		prev=temp;
		temp=temp->next;
	}

	if(temp==NULL){
		printf("value to be deleted not found\n");
	}
		

	prev->next=temp->next;
	free(temp);
}

void traverse(struct node* head){
	struct node* temp=head;
	while(temp!=NULL){
		printf(" %d  ->",temp->data);
		temp=temp->next;
	}
	printf(" NULL \n");
}		
	

void main(){
	struct node* head=NULL;
	insertAtBeg(&head,10);
	insertAtBeg(&head,30);

	traverse(head);
	insertAtEnd(&head,20);
	traverse(head);
	if(search(head,1)){
		printf("the elment is found\n");
	}else{
		printf("the element is not found\n");
	}

	deleteNode(&head,10);

	traverse(head);

}

File: python/data Strucutes/linkedlist.py
class Node:
    def __init__(self,data):
        self.data=data
        self.next=None

class LinkedList:
    def __init__(self):
        self.head=None

    def append(self,data):
        newNode=Node(data)

        if not self.head:
            self.head=newNode
            return
        curr=self.head

        while curr.next:
            curr=curr.next
        curr.next=newNode

    def prepend(self,data):
        newNode=Node(data)
        newNode.next=self.head
        self.head=newNode

    def delete(self,key):
        
        curr=self.head

        if curr and curr.data == key:
            self.head=curr.next
            return
        
        prev=None

        while curr and curr.data !=key:
            prev=curr
            curr=curr.next

        if curr:
            prev.next=curr.next


    def display(self):
        curr=self.head
        while curr:
            print(curr.data,end=" -> ")
            curr=curr.next
        print("None")



l1=LinkedList()
l1.append(1)
l1.append(2)
l1.prepend(10)
l1.display()

l1.delete(2)
l1.display()